Maintainer guidelines (#6)
ansible/community-docs: maintaining.rst

***********************
Maintaining collections
***********************

.. contents:: Topics

What is the collection maintenance
==================================

A collection is an Open Source project.

> If a community of any Open Source project can be logically divided into three groups: core members, contributors, and users, the collection maintainers are the core members of the collection.

Reviewer: Let's not use the term `core` to avoid confusion with the Core Team (who look after ansible-core).

Author: Good spot

Author: Changed, thanks

Who is a collection maintainer
------------------------------

A collection maintainer is a contributor trusted by the community who makes significant and regular contributions to the project and showed themselves as a specialist in the related area.

Collection maintainers has :ref:`extended permissions <Collection maintainers>` in the collection scope.

Collection maintainers satisfy the :ref:`requirements <Requirements for maintainers>`.

Requirements for maintainers
----------------------------

Maintainers act in accordance with `Ansible Code of Conduct <https://docs.ansible.com/ansible/latest/community/code_of_conduct.html>`_.

Maintainers (including candidates) have:

  - History of multiple contribution to a collection.
  - Excellent technical judgement in collection-related areas.
  - Responsiveness to mentions in issues and pull requests.
  - Responsiveness to issues and pull requests assigned to them.
  - Read these guidelines and the linked documents.
  - Subscribed to:

>     + a collection repository (the ``Watch`` button => ``All activity``),

Reviewer 2: ```suggestion
    + a collection repository (the ``Watch`` button → ``All activity``),
```

    + the ``Changes impacting collection contributors and maintainers`` `GitHub issue <https://github.com/ansible-collections/overview/issues/45>`_,
    + the `Bullhorn newsletter <https://github.com/ansible/community/issues/546>`_.

>   - Knowledge and intention to manage a collection performing the tasks listed in these guidelines. Maintainers can divided responsibilities between each other.

Reviewer 2: ```suggestion
  - Knowledge and intention to manage a collection performing the tasks listed in these guidelines. Maintainers can divide responsibilities between each other.
```

Author: @felixfontein thanks, committed the both

=============

We follow the `Ansible Code of Conduct <https://docs.ansible.com/ansible/latest/community/code_of_conduct.html>`_ in all interactions within the project.

Be informed
-----------

Good communication is vital for prosperity of the project.

Moreover, collection maintainers must be informed about important changes that impact all or many of the collections (for example, CI related) and act correspondingly to keep them up to date.

It is required for collection maintainers to be subscribed to the `"Changes impacting collection contributors and maintainers" GitHub issue <https://github.com/ansible-collections/overview/issues/45>`_ and the `Bullhorn newsletter <https://github.com/ansible/community/issues/546>`_.

Communication channels
----------------------

Collection contributors and maintainers communicate through:

  * the Bullhorn newsletter:

    + use the link in this `issue <https://github.com/ansible/community/issues/546>`_ to subscribe to the newsletter
    + if you have something important to announce (for example, releases made recently), put a comment in the issue
  * IRC channels such as ``#ansible-community``, ``#ansible-devel``, and dedicated ones

>   * mailing lists

Reviewer 3: ```suggestion
  * mailing lists such as [ansible-announce](https://groups.google.com/d/forum/ansible-announce) and [ansible-devel](https://groups.google.com/d/forum/ansible-devel)
```

.. _Backporting:

Backporting
===========

Collection maintainers backport merged pull requests to stable branches
following the `semantic versioning <https://semver.org/>`_ and release policies of the collections.

For more information about the process, refer to the `Backporting guidelines <https://docs.ansible.com/ansible/devel/community/development_process.html#backporting-merged-prs-in-ansible-core>`_.

For convenience, backporting can be implemented automatically using GitHub bots (for example, with the `Patchback app <https://github.com/apps/patchback>`_) and labeling like it is done in `community.general <https://github.com/ansible-collections/community.general>`_ and `community.network <https://github.com/ansible-collections/community.network>`_.

.. _Releasing:

Releasing
=========

Collection maintainers release all supported stable versions of the collections regularly,
provided that there have been enough changes merged to release.

Generally, releasing in the collections consists of:

  1. Planning and announcement.
  2. Generating a changelog.
  3. Creating a release git tag and pushing it.

>   4. Automatic publishing the release tarball on `Ansible Galaxy <https://galaxy.ansible.com/>`_ by Zuul.

Reviewer 3: ```suggestion
  4. Automatic publishing the release tarball on `Ansible Galaxy <https://galaxy.ansible.com/>`_ by [Zuul](https://dashboard.zuul.ansible.com/t/ansible/builds?pipeline=release).
```

  5. Final announcement.

For more information about the process, refer to the `Releasing guidelines <releasing.rst>`_.

.. _Expanding community:

Expanding community
===================

Increasing a number of active contributors and maintainers
----------------------------------------------------------

Maintainers are interested in increasing a number of active long-term contributors for a collection they maintain.

Contributors are reviewers, issue or pull request authors, testers, maintainers, and all other people who help develop the project.

> Every regular contributor once was a newcomer. Make the first experience as positive as possible to encourage the new people coming back.

Reviewer 3: ```suggestion
Every regular contributor was once a newcomer. Make the first experience as positive as possible to encourage the new people coming back.
```

The Ansible community hopes that you will find that maintaining a collection is as rewarding for you as having the collection content is for the wider community.

.. contents:: Topics:

Collection maintainership
=========================

What is the collection maintenance
----------------------------------

Ansible collections community can be logically divided into four major groups:

  1. Steering committee members (make decisions in scope of all the collections included in Ansible package and related areas).
  2. Maintainers.
  3. Contributors.
  4. Users.

Maintainers are people who are responsible for collection maintenance.

The collection maintenance consists of the regular activities listed in these guidelines.

In particular, collection maintainers:

  - keep README, development guidelines, and other general collection :ref:`documentation<Documentation>` relevant,
  - :ref:`review<Reviewing>` and :ref:`commit<Committing>` changes made by other contributors,
  - :ref:`backport<Backporting>` changes to stable branches,
  - address issues discovered to appropriate contributors,
  - :ref:`release<Releasing>` collections,

>   - watch that collections adhear the `Collection Requirements <https://github.com/ansible-collections/overview/blob/main/collection_requirements.rst>`_,

Author: haha:) committed, thanks!

  - increase a number of active contributors and maintainers by :ref:`building healthy community<Expanding community>` around collections.

Who is a collection maintainer
------------------------------

A collection maintainer is a contributor trusted by the community who makes significant and regular contributions to the project and showed themselves as a specialist in the related area.

Collection maintainers has :ref:`extended permissions<Collection maintainers>` in the collection scope.

Collection maintainers satisfy the :ref:`requirements<Requirements for maintainers>`.

.. _Requirements for maintainers:

Requirements for maintainers
----------------------------

Maintainers act in accordance with `Ansible Code of Conduct <https://docs.ansible.com/ansible/latest/community/code_of_conduct.html>`_.

Maintainers (including candidates) have:

  - History of multiple contribution to a collection.
  - Excellent technical judgement in collection-related areas.
  - Responsiveness to mentions in issues and pull requests.
  - Responsiveness to issues and pull requests assigned to them.
  - Read these guidelines and the linked documents.
  - Subscribed to:

>     + a collection repository (the ``Watch`` button → ``All activity``),

Author: yep:) I'll add a clarification, thanks! Committed this one

    + the ``Changes impacting collection contributors and maintainers`` `GitHub issue <https://github.com/ansible-collections/overview/issues/45>`_,
    + the `Bullhorn newsletter <https://github.com/ansible/community/issues/546>`_.
  - Knowledge and intention to manage a collection performing the tasks listed in these guidelines. Maintainers can divide responsibilities between each other.

How to become a maintainer
--------------------------

A person who is interested in becoming a maintainer and satisfies the :ref:`requirements<Requirements for maintainers>` may either self-nominate or be nominated by another maintainer.

> To nominate a candidate, create and issue under the `ansible/community <https://github.com/ansible/community>`_ repository.

Author: Yeah, would be nice to hear any ideas.
Also currently there are people with "Admin" permission from outside of the organization in our collections and they can add people ourselves (i've just recalled the c.mysql and c.proxysql cases).
Good question is if we should control this.
@gundalow what do you think?
I think we should have a policy of sort.

> To nominate a candidate, create and issue under the `ansible/community <https://github.com/ansible/community>`_ repository.

Reviewer: How about something simpler, and puts the discussion in the collection repo (where the right people are)
```suggestion
To nominate a candidate, please create a GitHub issue in the relevant collection repository repository.
```

> To nominate a candidate, create and issue under the `ansible/community <https://github.com/ansible/community>`_ repository.

Author: Sounds good. I added a sentence. @gundalow @felixfontein PTAL

.. _Communication:

Communication
=============

We follow the `Ansible Code of Conduct <https://docs.ansible.com/ansible/latest/community/code_of_conduct.html>`_ in all interactions within the project.

Be informed
-----------

Good communication is vital for prosperity of the project.

Moreover, collection maintainers must be informed about important changes that impact all or many of the collections (for example, CI related) and act correspondingly to keep them up to date.

It is required for collection maintainers to be subscribed to the ``Changes impacting collection contributors and maintainers`` `GitHub issue <https://github.com/ansible-collections/overview/issues/45>`_ and the `Bullhorn newsletter <https://github.com/ansible/community/issues/546>`_.

Communication channels
----------------------

Collection contributors and maintainers communicate through:

>   * the ``Changes impacting collection contributors and maintainers`` `GitHub issue <https://github.com/ansible-collections/overview/issues/45>`_

Author: i thought it's a kind of 1-to-many communication but agreed, will remove it (wanted to point this out one more time).

  * the Bullhorn newsletter:

    + use the link in this `issue <https://github.com/ansible/community/issues/546>`_ to subscribe to the newsletter
    + if you have something important to announce (for example, releases made recently), put a comment in the issue
  * IRC channels such as ``#ansible-community``, ``#ansible-devel``, and dedicated ones
  * mailing lists
  * collection pinboards, issues, and GitHub discussions in corresponding repositories
  * quarterly contributor summits
  * Ansible fests and local meetups

For more information about available IRC channels and mailing lists, refer to the `Ansible community documentation <https://docs.ansible.com/ansible/devel/community/communication.html>`_.

Weekly community IRC meetings
-----------------------------

> The important project-scale decisions are made by the community and the streeting committee at weekly IRC meetings in the ``#ansible-community`` IRC channel. See the `meeting schedule <https://github.com/ansible/community/blob/main/meetings/README.md#schedule>`_.

Reviewer 4: ```suggestion
The important project-scale decisions are made by the community and the steering committee at weekly IRC meetings in the ``#ansible-community`` IRC channel. See the `meeting schedule <https://github.com/ansible/community/blob/main/meetings/README.md#schedule>`_.
```

Reviewer 5: Maybe "Steering Committee" (capitalized)?

Author: Sounds good!

Module maintainers are contributors who are listed in ``.github/BOTMETA.yml``.

The scope can be any file (for example, a module or plugin), directory, or repository.

Because in most cases the scope is a module or group of modules, we call these contributors as module maintainers.

The collection bot notifies module maintainers when issues / pull requests related to files they maintain are created.

Module maintainers have the indirect commit right implemented through the `collection bot <https://github.com/ansible-community/collection_bot>`_.
When two module maintainers comment with the keywords ``shipit``, ``LGTM``, or ``+1`` a pull request
which changes a module they maintain, the collection bot will merge the pull request automatically.

For more information about the collection bot and its interface,
refer to the `Collection bot overview <https://github.com/ansible-community/collection_bot/blob/main/ISSUE_HELP.md>`_.

When a collection maintainer considers a contribution to a file significant enough
(it can be, for example, fixing a complex bug, adding a feature, providing regular reviews, and so on),

> they can offer the author to become a module maintainer, in other words, to add their GitHub login to ``.github/BOTMETA.yml``.

Reviewer 4: Not important though, feel free to ignore
```suggestion
they can offer the author to become a module maintainer, in other words, to add their GitHub account to ``.github/BOTMETA.yml``.
```

Author: LGTM, committed, thanks!

===========

Collection maintainers backport merged pull requests to stable branches
following the `semantic versioning <https://semver.org/>`_ and release policies of the collections.

For more information about the process, refer to the `Backporting guidelines <https://docs.ansible.com/ansible/devel/community/development_process.html#backporting-merged-prs-in-ansible-core>`_.

For convenience, backporting can be implemented automatically using GitHub bots (for example, with the `Patchback app <https://github.com/apps/patchback>`_) and labeling like it is done in `community.general <https://github.com/ansible-collections/community.general>`_ and `community.network <https://github.com/ansible-collections/community.network>`_.

.. _Releasing:

Releasing
=========

Collection maintainers release all supported stable versions of the collections regularly,
provided that there have been enough changes merged to release.

Generally, releasing in the collections consists of:

  1. Planning and announcement.
  2. Generating a changelog.
  3. Creating a release git tag and pushing it.
  4. Automatic publishing the release tarball on `Ansible Galaxy <https://galaxy.ansible.com/>`_ by Zuul.

>   5. Final announcement.

Reviewer 4: Maybe add a note about https://github.com/ansible/community/issues/546 here too?

Author: I tried to make it as general and short as possible in this document. The note about Bullhorn is already in `releasing_collections_with*` rst files here.

For more information about the process, refer to the `Releasing guidelines <releasing.rst>`_.

.. _Expanding community:

Expanding community
===================

Increasing a number of active contributors and maintainers
----------------------------------------------------------

Maintainers are interested in increasing a number of active long-term contributors for a collection they maintain.

> Contributors are reviewers, issue or pull request authors, testers, maintainer, and all other people who help develop the project.

Reviewer 4: ```suggestion
Contributors are reviewers, issue or pull request authors, testers, maintainers, and all other people who help develop the project.
```

Author: Good spot, committed, thanks!

> Every regular contributor once was a newcomer. Make the first experience as positive as possible to make the new people coming back.

Reviewer 4: ```suggestion
Every regular contributor once was a newcomer. Make the first experience as positive as possible to encourage the new people coming back.
```

Author: much better, thanks

Good development documentation makes contributors life much easier. Get feedback from new contributors if there were things they struggled with when working on their proposals and improve the documentation correspondingly.

Create the ``CONTRIBUTING`` file in your repository. In there, add a link to the `Quick-start guide <create_pr_quick_start_guide.rst>` as well as to other guidelines describing things specific to your collection.

Make contributors feel welcome. Greet and thank contributors impersonally in ``README`` and individually in their proposals.
Thank all participants after merging or closing a proposal.

Be responsive. Respond as quickly as possible. Even if you cannot review a proposal right now, greet and thank the author.

> Add and keep updated the ``CONTRIBUTORS`` file listing all the contributors including issue reporters and refer to it from your ``README``.

Reviewer 4: I'm not really sure about this, as keeping an eye on this for every new PR and new contributor would become hard.

Author: done

Do not fix trivial non-critical bugs yourself. Instead, mentor a person who would like to contribute.

> Mark issues with labels like ``easy_fix``, ``help_wanted``, and ``documentation``.

Reviewer 4: ```suggestion
Mark issues with labels like ``easy_fix``, ``waiting_on_contributor``, and ``documentation``.
```

Author: I like the idea. I think they aren't fully interchangeable, maybe both?

Reviewer 4: I haven't seen `help_wanted` tag on ansible/ansible or community repositories, also https://github.com/ansible-collections/community.general/pull/2602#discussion_r638678729.
Using one label for this type of issues would help both maintainers labeling and contributors finding these issues IMHO.

Author: persuaded:) done, thanks for explaining

Author: @aminvakil would you like to change `easy_fix` -> `easyfix` and `help_wanted` -> `waiting_on_contributor` in contributing.rst in this repo? Feel free to ping me, after raising a PR.

Reviewer 4: I'd be glad to, I will in a couple of hours.

They will let newcomers know where they can find easy wins.

When reviewing an issue, if applicable, ask the author whether they want to fix the issue themselves providing the link to the `Quick-start guide <create_pr_quick_start_guide.rst>`.

Adopt a zero-tolerance policy towards behavior violating `Ansible Code of Conduct <https://docs.ansible.com/ansible/latest/community/code_of_conduct.html>`_. Add information to ``README`` how people can complain referring to the `policy-violations Code of Conduct section<https://docs.ansible.com/ansible/latest/community/code_of_conduct.html#policy-violations>`_.

Announce that the project needs new contributors and maintainers through available communication channels.

Promote active contributors satisfying :ref:`requirements<Requirements for maintainers>` to maintainers. Revise contributors activity regularly.

> Create the ``MAINTAINERS`` file and keep it updated.

Reviewer 4: Do we have this file in any repository yet? I have only seen `commit-rights.md` file in c.g.

Author: I've recently added it to c.postgresq, c.proxysql, c.mysql and links from their READMEs to it. See https://github.com/ansible-collections/community.mysql. Though it doesn't mean I insist to have it here applied all across ansible-collections organization:)

***********************
Maintaining collections
***********************

> (Table content here)

Reviewer: ```suggestion
.. contents:: Topics```
